File: bst/bst.py
class BSTNode:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

    def __str__(self):
        return ' ' + str(self.data) + ' '


class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, root, data):
        new_node = BSTNode(data)

        if root is None:
            root = new_node
            return root

        elif data < root.data:
            root.left = self.insert(root.left, data)

        else:
            root.right = self.insert(root.right, data)
        return root


    def in_order_walk(self, root):
        result = ''
        if root is not None:
            result += str(root.data)
            result +=  ' ' + self.in_order_walk(root.left)

            result += ' ' + self.in_order_walk(root.right)
        return result

    def __str__(self):
        return self.in_order_walk(self.root)

if __name__ == '__main__':
    tree = BinarySearchTree()
    tree.root = tree.insert(tree.root, 10)

    tree.insert(tree.root, 20)


    tree.insert(tree.root, 15)

    print(tree)

File: bst/bst.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

struct node {
    int value;
    struct node* left;
    struct node* right;
};

struct node* insert(struct node*, int val);
void print(struct node*);

int main() {
    struct node* head;
    head = insert(head, 7);
    insert(head, 3);
    insert(head, 10);
    insert(head, -5);
    insert(head, 100);
    print(head);
}


struct node* insert(struct node* root, int val){
    
    if(root == NULL){
        // printf("here");
        struct node* new = (struct node *) malloc(sizeof(struct node));
        new->value = val;
        new->left = NULL;
        new->right = NULL;
        return new;
    }
    
    if(val >= root->value){
        if(root->right == NULL){
            struct node* new = (struct node *) malloc(sizeof(struct node));
            new->value = val;
            new->left = NULL;
            new->right = NULL;
            root->right = new;
            return new;
        }
        else{
            insert(root->right, val);
        }
    }
    else{
        if(root->left == NULL){
            struct node* new = (struct node *) malloc(sizeof(struct node));
            new->value = val;
            new->left = NULL;
            new->right = NULL;
            root->left = new;
            return new;
        }
        else{
            insert(root->left, val);
        }
    }

}


void print(struct node* root){
    if(root != NULL){
        print(root->left);
        printf("%d \n", root->value);
        print(root->right);
    }

}